spacy/pipeline/_parser_internals/nonproj.pyx

```diff
@@ -75,10 +76,19 @@ cdef bool _is_nonproj_arc(int tokenid, const vector[int]& heads) nogil:
 cdef bool _has_head_as_ancestor(int tokenid, int head, const vector[int]& heads) nogil:
     ancestor = tokenid
     cnt = 0
+    cdef cppset[int] seen_tokens
+    seen_tokens.insert(ancestor)
     while cnt < heads.size():
+        # Reached the head or a disconnected node
         if heads[ancestor] == head or heads[ancestor] < 0:
             return True
+        # Reached the root
+        if heads[ancestor] == ancestor:
+            return False
         ancestor = heads[ancestor]
+        if seen_tokens.count(ancestor):
```

Comment:
Sorry, we are using C++11, then use `find`, not `count` :).

Reply:
`find` seems to give an `iterator` https://github.com/cython/cython/blob/master/Cython/Includes/libcpp/unordered_set.pxd#L60. So it errors out saying that I cannot treat it as a `bool` in the `if`. I thought `count` is nice because i want it to be `0`.

Reply:
and this is faster?

Reply:
There shouldn't be any tangible difference in performance, but the `find` method is semantically more appropriate in this context.

Reply:
Okok, pushed it :-).

Reply:
> There shouldn't be any tangible difference in performance, but the `find` method is semantically more appropriate in this context.

Yeah, some standard libraries even implement `count` as literally

```
return haystack.find(needle) != haystack.end();
```

But it is, as @shadeMe says, about semantics. You want to know if the needle is in the haystack, not how many times. In some containers (e.g. `multiset` and `unordered_multiset`), `count` would also be slower.